### PR TITLE
Add get_metadata(), use_values option for get_features(), and dataset filtering script

### PR DESCRIPTION
**Pull request by tomingebretsencarlson**
### New: `get_metadata()` in `GroundTruthFeatures`
Returns a dict of per-point metadata aligned with the feature matrix rows:
- `cluster_id`: cluster index of each point
- `cluster_size`: number of points in the cluster
- `tree_height`: number of latent ancestors of each point

### New: `use_values` parameter in `get_features()`
`get_features(use_values=True)` stores the actual incremental latent value z_u 
(i.e. `node.value - parent.value`, or `node.value` for the root) instead of 1 
for each nonzero entry. Default is `False` to preserve existing behavior.
Updated `generate_data.py` to use `use_values=True` by default.

### Bug fix: `col_mask` after row filtering
`col_mask` previously used `.sum(axis=0) > 0` which incorrectly dropped columns 
with negative z_u values. Fixed to `(features != 0).sum(axis=0) > 0`.

### New: `scripts/filter_dataset.py`
Example script for filtering a saved dataset by cluster size, saving filtered 
X, y, features, and metadata.

### Tests
Added unit tests for `get_metadata()` and `get_features(use_values=True)`.

### CODE REVIEW

File: test_percolation_dataset.py [hunk 1]
```diff
@@ -339,6 +338,64 @@ def test_ground_truth_feature_matrix(self):
             self.assertEqual(set(non_zero_indices), set(expected_latent_indices),
                              f"Row {pidx} of ground truth feature matrix has incorrect non-zero entries")
             
+    def test_get_metadata(self):
+        """Test that get_metadata returns correct per-point metadata aligned with feature matrix rows."""
+        from collections import Counter
```
Comment (aribrill) [inline test_percolation_dataset.py]: Move to top-level imports

File: test_percolation_dataset.py [hunk 2]
```diff
@@ -339,6 +338,64 @@ def test_ground_truth_feature_matrix(self):
...
+        size = 1000
+        points, latents = self.dataset.construct(size=size)
+        gt_features = GroundTruthFeatures(points, latents)
+        metadata = gt_features.get_metadata()
+
+        self.assertIsInstance(metadata, dict)
+        self.assertListEqual(list(metadata.keys()), ['cluster_id', 'cluster_size', 'tree_height'])
+        for values in metadata.values():
+            self.assertEqual(len(values), size)
```
Comment (aribrill) [inline test_percolation_dataset.py]:
```suggestion
        for value in metadata.values():
            self.assertEqual(len(value), size)
```

File: percolation_dataset.py [hunk 3]
```diff
@@ -351,13 +351,34 @@ def __init__(self, points: List['Node'], latents: Dict[int, 'Node']):
...
+                - cluster_id: cluster index of each point
+                - cluster_size: number of points in each point's cluster
+                - tree_height: number of latent ancestors of each point
+        """
+
+        counts = Counter(p.cluster_idx for p in self.points)
+        return {
+            'cluster_id':   [p.cluster_idx for p in self.points],
+            'cluster_size': [counts[p.cluster_idx] for p in self.points],
+            'tree_height':  [len(self.pidx2lidx[i]) for i, _ in enumerate(self.points)],
```
Comment (aribrill) [inline percolation_dataset.py]: The name tree_height is ambiguous since there are several trees. Maybe latent_tree_height?

File: test_percolation_dataset.py [hunk 4]
```diff
@@ -339,6 +338,64 @@ def test_ground_truth_feature_matrix(self):
...
+        for i, point in enumerate(points):
+            self.assertEqual(metadata['cluster_id'][i], point.cluster_idx)
+            self.assertEqual(metadata['cluster_size'][i], counts[point.cluster_idx])
+            self.assertEqual(metadata['tree_height'][i], len(gt_features.pidx2lidx[i]))
+
+    def test_get_features_use_values(self):
+        """Test that get_features(use_values=True) stores correct z_u values."""
+        size = 100
+        points, latents, X, y = self.dataset.construct_embed(size=size, d=16)
+        gtf = GroundTruthFeatures(points, latents)
```
Comment (aribrill) [inline test_percolation_dataset.py]: Let's use consistent variable names throughout the code, if we can. Using gtf seems reasonable. Can you change this everywhere else to match?

File: test_percolation_dataset.py [hunk 5]
```diff
@@ -339,6 +338,64 @@ def test_ground_truth_feature_matrix(self):
...
+
+        X_binary = gtf.get_features(use_values=False)
+        X_values = gtf.get_features(use_values=True)
+
+        # Sparsity pattern must be identical
+        self.assertEqual((X_binary != 0).nnz, (X_values != 0).nnz)
```
Comment (aribrill) [inline test_percolation_dataset.py]: Can you more simply use X_binary.nnz, X_values.nnz here?

File: test_percolation_dataset.py [hunk 6]
```diff
@@ -339,6 +338,64 @@ def test_ground_truth_feature_matrix(self):
...
+            latent_node = gtf.latents[gtf.lidx2latent[lidx]]
+            if latent_node.parents:
+                parent = list(latent_node.parents)[0]
+                expected_z_u = latent_node.value - parent.value
+            else:
+                expected_z_u = latent_node.value
+            col = X_values.getcol(lidx)
+            nonzero_vals = col.data
+            if len(nonzero_vals) > 0:
+                np.testing.assert_allclose(nonzero_vals, expected_z_u,
```
Comment (aribrill) [inline test_percolation_dataset.py]: It might be better to use self.assertTrue(np.allclose()) to have consistent syntax throughout the module.

File: percolation_dataset.py [hunk 7]
```diff
@@ -369,7 +390,16 @@ def get_features(self, n_features: Optional[int] = None) -> sparse.csr_matrix:
         col_ind = []
 
         for lidx, pidx in islice(self.lidx2pidx.items(), n_features):
-            data.extend(np.ones_like(pidx))
+            if use_values:
+                latent_node = self.latents[self.lidx2latent[lidx]]
+                if latent_node.parents:
+                    parent = list(latent_node.parents)[0]
+                    z_u = latent_node.value - parent.value
```
Comment (aribrill) [inline percolation_dataset.py:397]: A possible concern is that the bottommost z_u will be confounded by the irreducible error terms for a model trained on the data. I think it is correct not to incorporate that into the ground-truth z_u values (so the approach taken here seems right to me), but it does mean that we should probably restrict our analyses to exclude the largest depths. I don't think this will matter much in practice anyway since the statistics are very limited at large depth anyway.

File: generate_data.py [hunk 8]
```diff
@@ -21,4 +22,8 @@
     sparse.save_npz(
         f"percolation_dataset_size{dataset_size}_dim{embedding_dimension}_seed{seed}_gt_features.npz",
         gt_features,
-    )
\ No newline at end of file
+    )
+    np.savez_compressed(
```
Comment (aribrill) [inline generate_data.py]: Probably you already looked into this, but it would be nice to combine the features and metadata into the same saved file to avoid having lots of similar files floating around. That just feels annoying and error-prone. It's not a huge deal though.

File: percolation_dataset.py [hunk 9]
```diff
@@ -351,13 +351,34 @@ def __init__(self, points: List['Node'], latents: Dict[int, 'Node']):
         self.n_samples = len(self.points)
         self.n_latents = len(self.lidx2pidx)
 
-    def get_features(self, n_features: Optional[int] = None) -> sparse.csr_matrix:
+
+    def get_metadata(self) -> dict:
```
Comment (aribrill) [inline percolation_dataset.py]: It seems to me that "metadata" is not quite the right term here. That usually connotes information like hyperparameter values, config settings, software versions, and timestamps (and in fact we should record these things). What's being stored here are still ground truth features, just ones that are at a higher level than the list of latents associated with each point. Maybe rename these two methods to get_latent_features and get_summary_features?

**Review by aribrill — COMMENTED** [review]
Looks good! I left some minor comments. In general, it's better to split distinct code changes into separate PRs when feasible (though I'm not always the best about this).

File: test_percolation_dataset.py (hunk 4 above)
Comment (tomingebretsencarlson) [reply test_percolation_dataset.py]: Changing gtf → gt_features to be consistent in terms of variables but also the names of the tests. Let me know if you still prefer gtf.

File: test_percolation_dataset.py (hunk 6 above)
Comment (tomingebretsencarlson) [reply test_percolation_dataset.py]: Changed, but I believe this removes debugging info if the this test breaks

File: percolation_dataset.py (hunk 7 above)
Comment (tomingebretsencarlson) [reply percolation_dataset.py:397]: Right, that is a good point!

File: percolation_dataset.py (hunk 9 above)
Comment (tomingebretsencarlson) [reply percolation_dataset.py]: Yes, I fully agree. Changing to your suggestions.

**Comment by tomingebretsencarlson** [comment]
> Looks good! I left some minor comments. In general, it's better to split distinct code changes into separate PRs when feasible (though I'm not always the best about this).

I agree, and will keep that in mind for future PRs. I’ve addressed the comments, updated the README to be compatible with the changes and updated generate_data.py for upstream compatibility.

File: generate_data.py (hunk 8 above)
Comment (tomingebretsencarlson) [reply generate_data.py]: Realized I can combine X, y with the cluster_id, cluster_size and latent_tree_height into one file. The change is done. 
Previously only thought about how to combine them into the sparse matrix.
